From: Philipp Zabel <p.zabel@pengutronix.de>
Date: Mon, 23 Feb 2015 16:45:23 +0100
Subject: [PATCH] media: coda: allow S_PARM on the CAPTURE side for backwards
 compatibility

gst-plugins-good before commit ce5c0b8f9271 ("v4l2object: set streamparm
for outputs that support it") did not support calling S_PARM on the
OUTPUT queue. Allow S_PARM on the CAPTURE queue instead for backwards
compatiblity with GStreamer versions older than 1.12.0.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>
---
 drivers/media/platform/coda/coda-common.c | 28 ++++++++++++++++++++++------
 1 file changed, 22 insertions(+), 6 deletions(-)

diff --git a/drivers/media/platform/coda/coda-common.c b/drivers/media/platform/coda/coda-common.c
index 17fe3abd95f3..e1e70d7ac9f7 100644
--- a/drivers/media/platform/coda/coda-common.c
+++ b/drivers/media/platform/coda/coda-common.c
@@ -1165,11 +1165,19 @@ static int coda_g_parm(struct file *file, void *fh, struct v4l2_streamparm *a)
 	struct coda_ctx *ctx = fh_to_ctx(fh);
 	struct v4l2_fract *tpf;
 
-	if (a->type != V4L2_BUF_TYPE_VIDEO_OUTPUT)
+	switch (a->type) {
+	case V4L2_BUF_TYPE_VIDEO_OUTPUT:
+		a->parm.output.capability = V4L2_CAP_TIMEPERFRAME;
+		tpf = &a->parm.output.timeperframe;
+		break;
+	case V4L2_BUF_TYPE_VIDEO_CAPTURE:
+		a->parm.capture.capability = V4L2_CAP_TIMEPERFRAME;
+		tpf = &a->parm.capture.timeperframe;
+		break;
+	default:
 		return -EINVAL;
+	}
 
-	a->parm.output.capability = V4L2_CAP_TIMEPERFRAME;
-	tpf = &a->parm.output.timeperframe;
 	tpf->denominator = ctx->params.framerate & CODA_FRATE_RES_MASK;
 	tpf->numerator = 1 + (ctx->params.framerate >>
 			      CODA_FRATE_DIV_OFFSET);
@@ -1248,11 +1256,19 @@ static int coda_s_parm(struct file *file, void *fh, struct v4l2_streamparm *a)
 	struct coda_ctx *ctx = fh_to_ctx(fh);
 	struct v4l2_fract *tpf;
 
-	if (a->type != V4L2_BUF_TYPE_VIDEO_OUTPUT)
+	switch (a->type) {
+	case V4L2_BUF_TYPE_VIDEO_OUTPUT:
+		a->parm.output.capability = V4L2_CAP_TIMEPERFRAME;
+		tpf = &a->parm.output.timeperframe;
+		break;
+	case V4L2_BUF_TYPE_VIDEO_CAPTURE:
+		a->parm.capture.capability = V4L2_CAP_TIMEPERFRAME;
+		tpf = &a->parm.capture.timeperframe;
+		break;
+	default:
 		return -EINVAL;
+	}
 
-	a->parm.output.capability = V4L2_CAP_TIMEPERFRAME;
-	tpf = &a->parm.output.timeperframe;
 	coda_approximate_timeperframe(tpf);
 	ctx->params.framerate = coda_timeperframe_to_frate(tpf);
 
